feat(tests): Add metrics in functional tests

eoger opened the pull request:
Still WIP, related to #1861. Would like some feedback on this :8ball:

zaach commented on app/scripts/lib/app-start.js:
```diff
@@ -596,6 +586,24 @@ function (
+
+    _isMetricsCollectionEnabled: function (sampleRate) {
+      return Math.random() <= sampleRate;
+    },
```
This method is a bit weird because it is nondeterministic; it may give you different results if called multiple times, whereas other methods with the `_is` prefix are deterministic. Maybe it fits better as a standalone helper function so we avoid the confusion if someone tries to call it multiple times like they would with other `_is` methods.

zaach commented on tests/functional/refreshes_metrics.js:
```diff
@@ -0,0 +1,51 @@
+        .get(require.toUrl(url))
+        .setFindTimeout(intern.config.pageLoadTimeout)
+
+        .findByCssSelector('form input.email')
```
We usually use the screen's header ID as the CSS selector, e.g. `fxa-signup-header`.

zaach commented on tests/functional/refreshes_metrics.js:
```diff
@@ -0,0 +1,51 @@
+    },
+
+    setup: function () {
+    },
```
You can omit these if they're empty.

zaach commented on tests/functional/refreshes_metrics.js:
```diff
@@ -0,0 +1,51 @@
+    setup: function () {
+    },
+
+    'signup page': function () {
```
The test name could be more descriptive, e.g. something like "refreshing signup logs a refresh event".

zaach commented on tests/functional/refreshes_metrics.js:
```diff
@@ -0,0 +1,51 @@
+
+    beforeEach: function () {
+
+    },
```
We may want to clear the browser state in this step so that local storage from previous tests doesn't interfere. Here's an example: https://github.com/mozilla/fxa-content-server/blob/master/tests/functional/change_password.js#L71

zaach replied:
Er, this feature uses `sessionStorage`, but just in case. :)

zaach commented:
This is a great first step toward #1861! In order to handle metrics testing more generally, we could do something like set `automatedBrowser=true` on all pages during functional testing so that metrics will be sent to localStorage. This will require us to revisit some functional tests that depend on that flag *not* being set. Or we can find another way to signal localStorage-metrics.

We can also add a test helper function to help assert that a set of events all occurred, and in the correct order. Then we'd have a pretty nice way to retrofit metrics assertions in all of our functional test flows. There may be some trickiness with tests that use multiple windows...

/cc @vladikoff @kparlante

zaach commented:
It's looks like there's one new line that's not covered by a unit test: https://coveralls.io/builds/2652996/source?filename=app%2Fscripts%2Flib%2Fapp-start.js#L593

You should be able to add a new `app-start` test that acts as if `automatedBrowser` is set and checks for storage metrics.

eoger commented:
Thanks @zaach for the very complete review, I updated the code.

> We can also add a test helper function to help assert that a set of events all occurred, and in the correct order.

Really good idea, I added a testAreEventsLogged which takes an ordered array of events as parameter.

zaach commented:
Looks like there was a formatting issues:
>     tests/functional/lib/helpers.js

>      line 602  col 53  Missing semicolon.

You can run `grunt lint` locally to find these before pushing to travis. :)

zaach commented:
Looks good! We'll wait until this email marketing stuff has landed before merging this.

jrgm commented:
> Looks good! We'll wait until this email marketing stuff has landed before merging this.

Is this a requirement for the spring launch?

zaach commented:
@jrgm No, this can safely wait for train 39.

zaach commented:
Ready to merge :+1:
